Kernel: Add support for the FAT32 filesystem

Undefine-Unity opened the pull request:
This commit adds read-only support for the FAT32 filesystem. It also includes support for long file names. As far as I know all the FAT32 and long file names patents have already expired, so we should be safe from legal issues. In the future I'm planning to implement write support and FAT12/FAT16 support, although I don't know if we even need it. Also FAT32 will be useful when it comes to UEFI booting, RPi booting and removable devices.

supercomputer7 commented:
As long as we won't have legal issues (I am not a lawyer) with FAT32 patents, I'm OK with this. I know in fact that exFAT requires a license from Microsoft to be implemented even in open source projects, but I guess that you did your checks to confirm that FAT32 will not be an issue for us.

filiphsps commented:
> As long as we won't have legal issues (I am not a lawyer) with FAT32 patents, I'm OK with this. I know in fact that exFAT requires a license from Microsoft to be implemented even in open source projects, but I guess that you did your checks to confirm that FAT32 will not be an issue for us.

From what I gather by reading patents we're still not allowed to implement exFAT without a license, see [US20080168029](https://patents.google.com/patent/US20080168029).
vFAT aka FAT32 with long file names expired in 2013 in the states (see [US20080168029](https://patents.google.com/patent/US20080168029)) and 2014 in Germany (see [EP0618540](https://data.epo.org/gpi/EP0618540B1-Common-name-space-for-long-and-short-filenames)).

vFAT is still however debated and IIRC the legal workaround that the likes of Linux uses is to read long file names separately thus circumventing the patent(s).

TL;DR FAT32 in itself is free to use, it's the extensions that are questionable.

(I am not a lawyer, and this is not legal advice. Please don't sue me).

Undefine-Unity commented:
First of all the EFI FAT32 specification gives permission for implementing FAT32 including long file names for purposes of EFI booting, which we will eventually use the driver for. Also all the relevant long file name / vFAT patents (see [US5579517](https://www.google.com/patents/US5579517), [US5745902](https://www.google.com/patents/US5745902), [US5758352](https://www.google.com/patents/US5758352) and [US6286013](https://www.google.com/patents/US6286013)) seem to have already expired.

Quaker762 commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,361 @@
+
+BlockBasedFileSystem::BlockIndex FATFS::first_block_of_cluster(u32 cluster) const
+{
+    return ((cluster - 2) * boot_record()->sectors_per_cluster) + m_first_data_sector;
```
Could you be a dear and make this a constant please :^)

Undefine-Unity replied:
Are you talking about the function?

Quaker762 replied:
No. The constant 2. This should be a named `constexpr`

Undefine-Unity replied:
This really isn't a constant, it's just a part of the calculation.

Quaker762 replied:
That's not true at all. The first two clusters are reserved special entries, hence `2` is something akin to `RESERVED_CLUSTER_COUNT`. This should be communicated instead of just sticking a random magic number into such a critical calculation.

Quaker762 commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,361 @@
+
+    u32 cluster = (((u32)m_entry.first_cluster_high) << 16) | m_entry.first_cluster_low;
+
+    if (cluster == 1)
```
Same here

Undefine-Unity replied:
What do you mean?

Quaker762 commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,361 @@
+{
+    dbgln_if(FAT_DEBUG, "FATFS: Reading inode {}: ize: {} offset: {}", identifier().index(), size, offset);
+
+    // FIXME: Read only the neaded blocks instead of the whole file
```
Typo: "neaded"->"needed"

Quaker762 commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,361 @@
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
+
+    for (auto& inode : children) {
+        // FIXME: For some reason using == doesn't work
```
What are you trying to use `operator==` on? The `inode` specifically? It doesn't look like you have `operator==()` implemented in your `FATInode` class, did you perhaps mean to implement that?

Undefine-Unity replied:
I tried comparing `inode->m_filename` and `name`, although it would always return false, later I found out that this is caused by the fact that `KString` adds a null byte at the end of the string. The thing is that using `strncmp` works just fine so I'm thinking of leaving this as it is.

timschumi replied:
That sounds like an issue with `KString` though (or rather, where that specific `KString` is constructed) if unexpected null bytes are appearing, as `KString` does store a length.

Quaker762 reviewed (changes requested):
Looks pretty good, few nits/questions :^)

timschumi commented on Kernel/FileSystem/FATFileSystem.h:
```diff
@@ -0,0 +1,157 @@
+    u8 entry_index;
+    u16 characters1[5];
+    u8 attribute;
+    u8 entry_type;
```
Is there an `enum` (or `enum`-like) type to go with this? (I do realize that the entry type seems to be currently unused by our code.)

Undefine-Unity replied:
The documentation says that it's always zero.

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,361 @@
+    } else {
+        StringBuilder filename;
+        for (auto& lfn_entry : lfn_entries) {
+
```
Empty line.

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,361 @@
+{
+    VERIFY(!m_filename);
+
+    // FIXME: Find a way to handle errors here instead of using MUST
```
This function could probably already be `ErrorOr<void>`, or at least I don't see a reason why it couldn't be. Of course, this only moves the "handling errors" part further down the line, but it's a first step.

Undefine-Unity replied:
The only place that function is called is the constructor which is not failable. I made the function `ErrorOr<void>` and moved that `FIXME` to the constructor, but I don't know if that's a good solution.

timschumi commented on Kernel/Debug.h.in:
```diff
@@ -87,6 +87,14 @@
 #cmakedefine01 EXT2_VERY_DEBUG
 #endif
 
+#ifndef FAT_DEBUG
+#cmakedefine01 FAT_DEBUG
+#endif
+
+#ifndef FAT_VERY_DEBUG
+#cmakedefine01 FAT_VERY_DEBUG
+#endif
+
```
Having two different log levels seems a bit over the top to me. The important parts of the initialization output (size, drive number, volume id, etc.) as well as some error messages could probably remain unguarded by a debug flag. The rest seems noisy enough (especially the lines on read/write operations) that it could be moved into `FAT_VERY_DEBUG` instead, which could then be the only debug flag, `FAT_DEBUG`.

Undefine-Unity replied:
While I agree that it's unnecessary to have two debug flags, but I still believe all the output should be guarded (except the invalid signature error).

timschumi commented on Kernel/FileSystem/FATFileSystem.h:
```diff
@@ -0,0 +1,157 @@
+private:
+    FATFS(OpenFileDescription&);
+
+    static constexpr u8 signature_1 = 0x28;
+    static constexpr u8 signature_2 = 0x29;
```
Even though it's two bytes that aren't interpreted together, for things like signature checking we would usually just read and compare them as a single `u16`.

Undefine-Unity replied:
They are not a one number, but two different possibilities.

timschumi commented on Kernel/FileSystem/FATFileSystem.h:
```diff
@@ -0,0 +1,157 @@
+    static constexpr u8 signature_1 = 0x28;
+    static constexpr u8 signature_2 = 0x29;
+
+    FAT32BootRecord const* boot_record() const { return (FAT32BootRecord const*)m_boot_record->data(); };
```
This should _probably_ be a C++ cast instead of a C cast, same with many other occurrences of casts in this file.

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,361 @@
+{
+    size_t length = max_length;
+
+    // HACK: size_t is unsigned, so we can't check if it's smaller than 0,
+    //       but when we go below zero the value turns really big
+    while (length <= max_length) {
```
That sounds like a prime opportunity for some weird values causing issues down the line (not that we probably survive until here anyways if something passes a length of `-1`).

Couldn't you start out with putting the `string` and `max_length` into a `StringView` at the very beginning, and then continue with cutting off bytes at the end? That would seem like a much safer alternative compared to working with raw pointers and offsets.

Undefine-Unity commented on Kernel/FileSystem/FATFileSystem.h:
```diff
@@ -0,0 +1,157 @@
+    u16 root_directory_entry_count;
+    u16 unused;
+    u8 media_descriptor_type;
+    u16 unused2;
```
There isn't any diffrence between `unused` and `reserved`, we aren't looking at any of these anyways. Changed the naming scheme to the same one used in the ISO9660 driver (start with `unused1` and increment).

Undefine-Unity commented on Kernel/FileSystem/FATFileSystem.h:
```diff
@@ -0,0 +1,157 @@
+struct [[gnu::packed]] FATLongFileNameEntry {
+    u8 entry_index;
+    u16 characters1[5];
+    u8 attribute;
```
Long file name entries have their unique attribute and the value is always the same. I changed that to `attributes` just for consistency.

Undefine-Unity commented on Kernel/FileSystem/FATFileSystem.h:
```diff
@@ -0,0 +1,157 @@
+struct [[gnu::packed]] FATLongFileNameEntry {
+    u8 entry_index;
+    u16 characters1[5];
+    u8 attribute;
```
As I said below the value is always the same, but it is one of the possible attributes so I changed that.

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,358 @@
+    dbgln_if(FAT_DEBUG, "FATFS: Reading inode {}: ize: {} offset: {}", identifier().index(), size, offset);
+
+    // FIXME: Read only the needed blocks instead of the whole file
+    auto blocks = TRY(const_cast<FATInode&>(*this).read_block_list());
```
That looks very weird. Maybe move `generate_block_list` somewhere else and make `read_block_list` `const` as well.

Undefine-Unity replied:
There isn't an easy way to make `read_block_list` `const`. It uses some read functions that are not marked `const` (maybe they should be?) and also I think we should only compute the block list if we need in place like here, because let's say you want to list a directory with a very large amount of files, it would be pointless to run this function for every one of the files because we only about their file name.

timschumi reviewed (changes requested):
Another round with what I originally missed.

Undefine-Unity commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,358 @@
+}
+
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(u8 const* initial_string, size_t max_length, u8 byte) const
```
It can't be made `char` because long file name entries are terminated by `0xFF` and the compiler would complain if we passed it as `char`.

ADKaster commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,352 @@
+
+    for (auto& inode : children) {
+        if (strncmp(reinterpret_cast<char const*>(inode->m_filename->bytes().data()), reinterpret_cast<char const*>(name.bytes().data()), name.length()) == 0) {
+            return TRY(adopt_nonnull_lock_ref_or_enomem(inode.leak_ref()));
```
This looks very suspicious.

Why are we leaking a reference? 

This makes it look like we have some places holding a RefPtr to these FATInode objects, and other places holding a Locking RefPtr. That seems very unsafe. If the parent INode classes expect things to be returned as locking RefPtrs, we should probably use that everywhere to not trip over ourselves later and fix the locking story at a higher level.

ADKaster commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,352 @@
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
+
+    for (auto& inode : children) {
+        if (strncmp(reinterpret_cast<char const*>(inode->m_filename->bytes().data()), reinterpret_cast<char const*>(name.bytes().data()), name.length()) == 0) {
```
This should just be `if (inode->m_filename->view() == name)`

It's very hard to read when you reach into both of these objects that are equality comparable to each other and reinterpret cast their bytes and use a C function.

ADKaster commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,352 @@
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
```
I don't understand why we have to allocate a Vector of RefPtr to every child of this INode just to find the specific INode that has the passed-in name.

Can't traverse() take a `Function<bool(FATINode&)>` similar to other traverse/find functions in the system?

`RefPtr<FatINode> found_item = traverse_and_find([name](auto& child) {  return (child.m_filename.view() == name);' });`

ADKaster commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,352 @@
+            dbgln_if(FAT_DEBUG, "FATFS: Found unused entry");
+        } else if (entry->attributes == FATAttributes::LongFileName) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found LFN entry");
+            lfn_entries.append(*reinterpret_cast<FATLongFileNameEntry*>(entry));
```
this can fail to allocate. as can the append on inodes below.

ADKaster commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,352 @@
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(StringView string, u8 fill_byte)
+{
+    while (string.length() > 0) {
```
It's very hard to figure out what this algorithm is trying to do.

It *looks* like you're doing some kind of `find_last_not_of`?

```
Optional<size_t> find_last_not_of(StringView haystack, char needle)
{
    for (size_t i = haystack.length(); i > 0; --i) {
        if (haystack[i - 1] != needle)
            return i - 1;
    }
    return {};
}


StringView FATInode::byte_terminated_string(StringView string, u8 fill_byte)
{
    if (auto index = find_last_not_of(string, fill_byte); index.has_value())
         return string.substring_view(0, index.value();
    return string;
}
```

ADKaster commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,352 @@
+    Vector<RefPtr<FATInode>> children = TRY(const_cast<FATInode&>(*this).traverse());
+
+    for (auto const& inode : children) {
+        TRY(callback({ inode->m_filename->view(), inode->identifier(), static_cast<u8>(inode->m_entry.attributes) }));
```
Same comment here.

We don't need to create a vector of RefPtrs just to invoke a callback on each item. we can call the callback while traversing.

ADKaster commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,352 @@
+    auto fat_sector = TRY(KBuffer::try_create_with_size("FATFS: FAT read buffer"sv, fs().m_logical_block_size));
+    auto fat_sector_buffer = UserOrKernelBuffer::for_kernel_buffer(fat_sector->data());
+
+    while (cluster < 0x0FFFFFF8) {
```
What do the magic numbers 0x0FFFFFF8 and 0x0FFFFFFF mean? Please give names to your constants. Similar to Quaker's comment on why the first two clusters are reserved.

If we have to read the FAT32 specification to understand the code, it is not descriptive enough.

bgianfo replied:
This seems like a valid critique of the code,  commenting constants isn't the serenity style. We name constants everywhere else in the codebase. This file system isn't special in that regard.

Please address this feedback.

Quaker762 replied:
> I don't think that making constants for numbers used in only one place is a good idea. Instead I decided to add comments describing what do these numbers mean.

It significantly improves readability to someone that wants to poke around later on, and, as @bgianfo  has said, is in the Serenity style which dictates that there are to be no magic numbers, least of all not in the Kernel of all places.

Please fix this.

Undefine-Unity replied:
Done.

ADKaster reviewed (commented):
At the risk of adding too many cooks to the kitchen, here's my review...

I think the string iteration function has already had a few rounds of refactor, so feel free to ignore that if i'm talking nonsense.

I think that traversal is a bit unoptimized, but that can probably be addressed later.

The conversion of RefPtr to LockedRefPtr does look like a real thread safety  issue though...

Undefine-Unity commented:
I think I addressed all of the constants in one way or another. In the `fat_date_time` function I decided to add comments describing how times and dates are stored in FAT32, I think those comments describe what the code is doing in a way better way than having constants for all the masks and bitshifts. I replaced the rest of the magic numbers with constant variables.

timschumi commented on Kernel/FileSystem/FATFileSystem.h:
```diff
@@ -0,0 +1,177 @@
+
+    static constexpr u8 lfn_entry_text_termination = 0xFF;
+
+    static constexpr u16 first_fat_year_offset = 10; // 1980 (first possible year that can be stored as a fat date) - 1970 (first unix year)
```
Comment should be above the code that it comments, not behind it (unless the comment is _very_ short, which it isn't).

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,357 @@
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    // Date format:
+    // - Year  7 bits
+    // - Month 4 bits
+    // - Day   5 bits
```
This information might be better stored in a proper bit-field. That makes the code easier to read, and it saves us from potential mistakes while shifting/masking. Same for the "Time format" below.

Undefine-Unity replied:
Is there an pre-existing structure / class that I could use for that?

timschumi replied:
It would be something similar to [this](https://github.com/SerenityOS/serenity/blob/master/AK/FPControl.h#L41-L62), but obviously purpose-fit to the problem at hand.

I'd suggest the following (and a respective second definition for `FATPackedTime`):

```
union FATPackedDate {
    u16 value;
    struct {
        u16 year : 7;
        u16 month : 4;
        u16 day : 5; 
    };
};
static_assert(sizeof(FATPackedDate) == 16);
```

That way, you can access `packed_value.value` to read/write the raw value that the filesystem actually stores, and `packed_value.year`, `.month` and `.day` to read/write the parsed values.

This is also compatible with members inside structs, so you could define the various date and time members as that type directly (i.e. `FATPackedDate last_accessed_date` inside of `FATEntry`);

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,357 @@
+    VERIFY_NOT_REACHED();
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
```
This should even be a `static` helper function, right? Not just `const`.

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,357 @@
+    time_t minute = (time >> 5) & 0b1111111;
+    time_t second = (time & 0b11111) * 2;
+
+    return second + minute * 60 + hour * 3600 + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
```
Is assuming 30 days in a month and 365 days in a year correct?

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,357 @@
+    // - Month 4 bits
+    // - Day   5 bits
+
+    time_t year = (date >> 9) + first_fat_year_offset;
```
In my opinion, this obstructs the clarity of the purpose of the offset quite a bit, even if it's appropriately explained with a comment where the constant is defined.

I'd rather do `time_t year = first_fat_year + (date >> 9);` here (where `first_fat_year` is 1980) and subtract the value 1970 in the formula at the end of the function (as that is the only part of the function that is actually concerned with making it a UNIX timestamp).

timschumi commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,357 @@
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    [[maybe_unused]] auto inode = TRY(const_cast<FATInode&>(*this).traverse([&callback](auto inode) -> ErrorOr<bool> {
```
You could probably just drop the `[[maybe_unused]] auto inode = ` completely here.

Undefine-Unity replied:
I can't, the compiler complains if I drop that.

timschumi reviewed (changes requested):
Looks like it's coming together quite nicely. :^)

Undefine-Unity commented on Kernel/FileSystem/FATFileSystem.cpp:
```diff
@@ -0,0 +1,357 @@
+    TRY(BlockBasedFileSystem::initialize());
+
+    FATEntry root_entry {};
+    root_entry.first_cluster_low = FATInode::root_inode_index;
```
Yes they do, that's a part of the `FATEntry` struct which we do not have control of, because it's a part of the spec. I have added a helper function to read it.

Undefine-Unity replied:
I ended up getting rid of the `root_inode_index` constant and replaced it with the value that would be actually used later, now both `first_cluster_low` and `first_cluster_high` are being written when creating the root inode.

timschumi reviewed (changes requested):
One last formality remains as far as I can tell, after that the PR gets a green check mark from me. :^)

timschumi reviewed (approved):
LGTM, and seems to work fine.
